src/main/webapp/script.js

```diff
@@ -217,54 +217,105 @@ function getGraphDisplay(graphNodes, graphEdges) {
-    }
-  };
-  
+  document.getElementById('reset').onclick = function(){ resetElements(cy) };
+
+  document.getElementById('search-button').onclick = function() { search(cy, "node", searchNode) };
+
+  document.getElementById('search-token-button').onclick = function() { search(cy, "token", searchToken) };
 }
 
 /**
- * Zooms in on specific node
+ * Searches based on type (node or token)
  */
-function searchNode(cy, query) {
-  // reset nodes to default color
-  cy.nodes().forEach(node => {
-    node.style('background-color', 'blue');
-    node.style('opacity', '1')
-  });
-  const target = findNodeInGraph(cy, query);
-  if (target) {
-    cy.nodes().forEach(node => node.style('opacity', '0.25'));
-    target.style('background-color', 'olive');
-    target.style('opacity', '1');
-    cy.fit(target, 50);
-    return true;
+function search(cy, type, searchFunction) {
+  resetElements(cy);
+  let errorText;
+  let query = document.getElementById(type + '-search').value;
+  let result = searchFunction(cy, query);
+  if (query == "" || result) {
+    errorText = "";
   } else {
-    // fits all nodes on screen
-    cy.fit(cy.nodes(), 50);
-    return false;
+    errorText = type + " does not exist.";
   }
+  document.getElementById(type + '-error').innerText = errorText;
+  return result;
 }
 
 /**
- * Finds element in cy graph by id
+ * Finds specific node and zooms in
  */
-function findNodeInGraph(cy, id) {
-  if (id.length != 0) {
-    const target = cy.$('#'+id);
+function searchNode(cy, query) {
+  if (query) {
+    const target = cy.$id(query);
     if (target.length != 0) {
+      highlightElements(cy, target);
+      cy.fit(target, 50);
       return target;
     }
   }
-  return null;
+  //return false;
+}
+/**
+ * Constructs list of nodes that contain specified token
+ * and zooms in
+ */
+function searchToken(cy, query) {
+  let target = cy.collection();
+  cy.nodes().forEach(node => {
+    if (node.data().tokens.includes(query)) {
+      target = target.add(cy.$id(node.data().id));
+    }
+  });
+  if (target.length > 0) {
+    let showNode = target[0][0];
+    showNode.tip.show();
+    highlightElements(cy, target);
+    return target;
+  }
+  //return false;
+}
+
+/**
+ * Highlights collection of nodes and edges
+ */
+function highlightElements(cy, target) {
+  // set all nodes to background state
+  cy.nodes().forEach(node => node.style('opacity', '0.25'));
+
+  // highlight desired nodes
+  target.forEach(node => {
+    node.style('border-width', '4px');
+    node.style('opacity', '1');
+  });
+  cy.fit(target[0], 50);
+  document.getElementById('num-selected').innerText = "Number of nodes selected: " + target.length;
+
+  // highlight adjacent edges
+  target.connectedEdges().forEach(edge => {
+    edge.style('line-color', 'black');
+    edge.style('target-arrow-color', 'black');
+  });
+}
+
+/**
+ * Resets elements to default state
+ */
+function resetElements(cy) {
+  cy.nodes().forEach(node => {
```

Comment:
We might not want to do this for all nodes, for example if combined with the mutation highlighting feature. Could we just reset the elements we modified?

Reply:
sure but the mutations only change the color of the nodes right? I'm not changing the color here at all so the mutated nodes colors should not change